rpcclient/rawtransactions.go

```diff
@@ -882,3 +883,117 @@ func (c *Client) DecodeScriptAsync(serializedScript []byte) FutureDecodeScriptRe
 	return c.DecodeScriptAsync(serializedScript).Receive()
 }
+
+// FutureTestMempoolAcceptResult is a future promise to deliver the result
+// of a TestMempoolAccept RPC invocation (or an applicable error).
+type FutureTestMempoolAcceptResult chan *Response
+
+// Receive waits for the Response promised by the future and returns the
+// response from TestMempoolAccept.
+func (r FutureTestMempoolAcceptResult) Receive() (
+	[]*btcjson.TestMempoolAcceptResult, error) {
+
+	response, err := ReceiveFuture(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal as an array of TestMempoolAcceptResult items.
+	var results []*btcjson.TestMempoolAcceptResult
+
+	err = json.Unmarshal(response, &results)
+	if err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+// TestMempoolAcceptAsync returns an instance of a type that can be used to get
+// the result of the RPC at some future time by invoking the Receive function
+// on the returned instance.
+//
+// See TestMempoolAccept for the blocking version and more details.
+func (c *Client) TestMempoolAcceptAsync(txns []*wire.MsgTx,
+	maxFeeRate float64) FutureTestMempoolAcceptResult {
+
+	// Due to differences in the testmempoolaccept API for different
+	// backends, we'll need to inspect our version and construct the
+	// appropriate request.
+	version, err := c.BackendVersion()
+	if err != nil {
+		return newFutureError(err)
+	}
+
+	log.Debugf("TestMempoolAcceptAsync: backend version %s", version)
+
+	// Exit early if the version is below 25.0.0.
+	if version < BitcoindPost25 {
+		err := fmt.Errorf("%w: %v", ErrBitcoindVersion, version)
+		return newFutureError(err)
+	}
+
+	// The maximum number of transactions allowed is 25.
+	if len(txns) > 25 {
+		err := fmt.Errorf("%w: too many transactions provided",
+			ErrInvalidParam)
+		return newFutureError(err)
+	}
+
+	// Exit early if an empty array of transactions is provided.
+	if len(txns) == 0 {
+		err := fmt.Errorf("%w: no transactions provided",
+			ErrInvalidParam)
+		return newFutureError(err)
+	}
+
+	// Iterate all the transactions and turn them into hex strings.
+	rawTxns := make([]string, 0, len(txns))
+	for _, tx := range txns {
+		// Serialize the transaction and convert to hex string.
+		buf := bytes.NewBuffer(make([]byte, 0, tx.SerializeSize()))
+
+		// TODO(yy): add similar checks found in `BtcDecode` to
+		// `BtcEncode` - atm it just serializes bytes without any
+		// bitcoin-specific checks.
+		if err := tx.Serialize(buf); err != nil {
+			err = fmt.Errorf("%w: %v", ErrInvalidParam, err)
+			return newFutureError(err)
+		}
+
+		rawTx := hex.EncodeToString(buf.Bytes())
+		rawTxns = append(rawTxns, rawTx)
+
+		// Sanity check the provided tx is valid, which can be removed
+		// once we have similar checks added in `BtcEncode`.
+		//
+		// NOTE: must be performed after buf.Bytes is copied above.
+		//
+		// TODO(yy): remove it once the above TODO is addressed.
+		if err := tx.Deserialize(buf); err != nil {
```

Comment:
I think I'm missing something here: we just serialized it above, why wouldn't it deserialize here?

Reply:
this is because `tx.Serialize` doesn't do any check to make sure it's a valid tx, one can pass an empty `*wire.MsgTx` to it and it will decode `[0 0 0 0 0 0 0 0 0 0]` into the buffer.